ensure extension triggers are only run by the package that satified them

vtjnash opened the pull request:
Oscar had noticed a problem where loading an extension might trigger loading another package, which might re-trigger attempting to load the same extension. And then that causes a deadlock from waiting for the extension to finish loading (it appears to be a recursive import triggered from multiple places). Instead alter the representation, to be similar to a semaphore, so that it will be loaded only exactly by the final package that satisfied all dependencies for it.

This approach could still encounter an issue if the user imports a package (C) which it does not explicitly list as a dependency for extension. But it is unclear to me that we actually want to solve that, since it weakens and delays the premise that Bext is available shortly after A and B are both loaded.

\# module C; using A, B; end;; module A; end;; module B; end;; module Bext; using C; end
\# using C, Bext / A / B
starts C -> requires A, B to load
loads A -> defines Bext (extends B, but tries to also require C) loads B -> loads Bext (which waits for C -> deadlock!)
finish C -> now safe to load Bext

While using this order would have been fine.
\# using A, B, Bext / C
loads A -> defines Bext (extends B, but tries to also require C) loads B -> starts Bext
loads C
finish Bext

vtjnash committed:
ensure extension triggers are only run by the package that satified them

Oscar had noticed a problem where loading an extension might trigger
loading another package, which might re-trigger attempting to load the
same extension. And then that causes a deadlock from waiting for the
extension to finish loading (it appears to be a recursive import
triggered from multiple places). Instead alter the representation, to be
similar to a semaphore, so that it will be loaded only exactly by the
final package that satisfied all dependencies for it.

This approach could still encounter an issue if the user imports a
package (C) which it does not explicitly list as a dependency for
extension. But it is unclear to me that we actually want to solve that,
since it weakens and delays the premise that Bext is available shortly
after A and B are both loaded.

\# module C; using A, B; end;; module A; end;; module B; end;; module Bext; using C; end
\# using C, Bext / A / B
starts C -> requires A, B to load
loads A -> defines Bext (extends B, but tries to also require C)
loads B -> loads Bext (which waits for C -> deadlock!)
finish C -> now safe to load Bext

While using this order would have been fine.
\# using A, B, Bext / C
loads A -> defines Bext (extends B, but tries to also require C)
loads B -> starts Bext
loads C
finish Bext

vtjnash commented:
Oscar helped discuss offline that this deadlock case (mentioned in the comments) is actually potentially a severe flaw here right now in EXT_DORMATORY. We have unintentionally created a situation where user might easily accidentally create a loading graph that deadlocks unpredictably when loading it. In practice, that means, in addition to this PR being sensible, that we may want to forbid extension packages from loading any new packages (unless we can prove they do not depend on any of the triggers?). It should thus potentiall be forbidden from accessing any package that was not explicitly listed as one of its triggers (any others it wants transitively it could access via `using Parent.Other` if desired).

KristofferC commented:
> It should thus potentiall be forbidden from accessing any package that was not explicitly listed as one of its triggers (any others it wants transitively it could access via using Parent.Other if desired).

Originally, I had that as the implementation and you went via `Parent.Other` to get dependencies. That however excluded the case where you wanted to load a package in an extension that was not a trigger nor was it loaded into `Parent`. This seemed to be a quite common desire so I allowed that. 

The code that allows that is this:

https://github.com/JuliaLang/julia/blob/c4fd8a48383235e818c44231d7507168c27f393b/base/loading.jl#L824-L825

KristofferC commented:
This also seems subsume https://github.com/JuliaLang/julia/pull/48509 and fix #48498?

vtjnash commented:
There are a couple problems we observed, and this only really addresses one of them.

This addresses the case where multiple extensions need to be loaded. Previously each extension that loaded would trigger a new walk through the list, and then load the next extension in sequence. This is not strictly a problem, but is a bit awkward as it means the stack keeps getting longer and complicates the locking order of the loaded packages in some cases.

> Originally, I had that as the implementation and you went via Parent.Other to get dependencies. That however excluded the case where you wanted to load a package in an extension that was not a trigger nor was it loaded into Parent. This seemed to be a quite common desire so I allowed that.

Secondly, we observed that this should never have been allowed. It would be very useful, if it was valid. But it creates a cycle in the loading graph, which leads to unpredictable deadlocks, and causing the extension to sometimes hang (v1.9) or error (master today). It turns out that we must forbid this (for now), to prevent such unreliable behavior from catching users unawares and making PkgEval unreliable. We can re-evaluate later if we want to design a solution for it to reallow it later. Of note, it also must be forbidden (aka strongly discouraged) from loading those extra, unexpected packages via other mechanisms (e.g. during `__init__` or Requires.jl too) for the same reason that they will sometimes cause it to deadlock or error unpredictably.

vtjnash commented:
> problem in the previous code was the use of Base.loaded_modules

FWIW, this was not a problem. I still check both of those lists, I just moved where it checked them, and implemented the TODO comment about making this more efficient than a repeated scan of the list.

KristofferC commented:
> But it creates a cycle in the loading graph, which leads to unpredictable deadlocks, and causing the extension to sometimes hang (v1.9) or error (master today).

Would moving the extension into the parent package create the same cycle?

vtjnash commented:
Commonly, yes. The same extension being loaded explicitly by the parent does suffer the same problem (the parent depends on weakdep depends on external which depends on the parent), but not if the same content is moved into another context downstream of the external dependency.

KristofferC commented:
> FWIW, this was not a problem.

But the previous code only loaded the extension when all the triggers were in `Base.loaded_modules` (so only the last one should trigger the loading). So what was the issue then?

vtjnash commented:
Yeah, we realized later this is not really the source of the problem with deadlock that we had encountered (though it was a possible cause), but I didn't go back to update the PR text. This is only a change to make it so that only exactly that package will run all of its triggers, and not any other package (including weakdeps) that happen to finish loading around the same time.

KristofferC commented:
Something seems strange here, when an extension fails to load, it tries to load it again. I introduced a syntax error in https://github.com/KristofferC/PGFPlotsX.jl/blob/master/ext/ColorsExt.jl and I get:

```julia
julia> using PGFPlotsX

julia> using Colors
[ Info: Precompiling ColorsExt [283d1826-985b-5544-82b8-7fd9aa83b823]
ERROR: LoadError: UndefVarError: `eeh` not defined
Stacktrace:
 [1] top-level scope
   @ ~/JuliaPkgs/PGFPlotsX.jl/ext/ColorsExt.jl:7
 [2] include
   @ ./Base.jl:456 [inlined]
 [3] include_package_for_output(pkg::Base.PkgId, input::String, depot_path::Vector{String}, dl_load_path::Vector{String}, load_path::Vector{String}, concrete_deps::Vector{Pair{Base.PkgId, UInt128}}, source::Nothing)
   @ Base ./loading.jl:2031
 [4] top-level scope
   @ stdin:2
in expression starting at /Users/kristoffercarlsson/JuliaPkgs/PGFPlotsX.jl/ext/ColorsExt.jl:1
in expression starting at stdin:2
┌ Error: Error during loading of extension ColorsExt of PGFPlotsX
└ @ Base loading.jl:1187
[ Info: Precompiling ColorsExt [283d1826-985b-5544-82b8-7fd9aa83b823]
ERROR: LoadError: UndefVarError: `eeh` not defined
Stacktrace:
 [1] top-level scope
   @ ~/JuliaPkgs/PGFPlotsX.jl/ext/ColorsExt.jl:7
 [2] include
   @ ./Base.jl:456 [inlined]
 [3] include_package_for_output(pkg::Base.PkgId, input::String, depot_path::Vector{String}, dl_load_path::Vector{String}, load_path::Vector{String}, concrete_deps::Vector{Pair{Base.PkgId, UInt128}}, source::Nothing)
   @ Base ./loading.jl:2031
 [4] top-level scope
   @ stdin:2
in expression starting at /Users/kristoffercarlsson/JuliaPkgs/PGFPlotsX.jl/ext/ColorsExt.jl:1
in expression starting at stdin:2
┌ Error: Error during loading of extension ColorsExt of PGFPlotsX
└ @ Base loading.jl:1187
```

with this PR (and using https://github.com/JuliaLang/julia/pull/48550 to make stuff a bit shorter).

KristofferC commented on base/loading.jl:
```diff
@@ -1159,59 +1158,84 @@ end
+        else
+            succeeded = true
+        end
+        if extid.ntriggers == 0
```
This should also check that the extension is not in `EXT_DORMITORY_FAILED`?

KristofferC replied:
Maybe something like

```diff
diff --git a/base/loading.jl b/base/loading.jl
index 25c89c31ca..e57e2e530a 100644
--- a/base/loading.jl
+++ b/base/loading.jl
@@ -1105,7 +1105,7 @@ mutable struct ExtensionId
 end
 
 const EXT_DORMITORY = Dict{PkgId,Vector{ExtensionId}}()
-const EXT_DORMITORY_FAILED = ExtensionId[]
+const EXT_DORMITORY_FAILED = Dict{PkgId,ExtensionId}()
 
 function insert_extension_triggers(pkg::PkgId)
     pkg.uuid === nothing && return
@@ -1229,11 +1229,11 @@ function run_extension_callbacks(pkgid::PkgId)
         else
             succeeded = true
         end
-        if extid.ntriggers == 0
+        if extid.ntriggers == 0 && !haskey(EXT_DORMITORY_FAILED, extid.id)
             # actually load extid, now that all dependencies are met,
             # and record the result
             succeeded = succeeded && run_extension_callbacks(extid)
-            succeeded || push!(EXT_DORMITORY_FAILED, extid)
+            succeeded || (EXT_DORMITORY_FAILED[extid.id] = extid)
         end
     end
     nothing
@@ -1253,10 +1253,10 @@ function retry_load_extensions()
     # so this can still mutate the list to drop successful ones
     failed = copy(EXT_DORMITORY_FAILED)
     empty!(EXT_DORMITORY_FAILED)
-    filter!(failed) do extid
+    filter!(failed) do (_, extid)
         return !run_extension_callbacks(extid)
     end
-    prepend!(EXT_DORMITORY_FAILED, failed)
+    merge!(EXT_DORMITORY_FAILED, failed)
     end
     nothing
 end
```

vtjnash replied:
That could be an interesting change, though I don't think it is strictly related to this PR. It only changes the case where something failed failed to load, while previously I was not intending to prevent a package from registering the same extension multiple times, possibly with different trigger sets.

vtjnash commented:
Oh, I think that is actually correct behavior, but only because of a different bug. The original PR for weakdeps usually puts 2 copies of the weakdep into the array. But the try-catch was in the wrong place, so it would initially skip the second one and then come back to all of them the next time you loaded any package to try again. I meant to file an issue, but forgot.

KristofferC commented:
> The original PR for weakdeps usually puts 2 copies of the weakdep into the array.

I actually think https://github.com/JuliaLang/julia/pull/48352 was what made this happen.

vtjnash commented:
I guess that might sometimes, but I think that PR is correct. I was referring to the presence of a call to that `insert_extension_triggers` function from both tryrequire and require however seeming that it may potentially insert duplicates.

KristofferC committed:
improve error message when an extension fails to load (#48550)

KristofferC committed:
only add extensions from one environment (#48558)

(cherry picked from commit b2adb8548d7f2a38dc73ea2de1be271e688a545c)

Krastanov commented:
This pull request seems to have caused a drop in coverage of 5 percentage points.
Or maybe I am simply misunderstanding the codecov report? 

https://app.codecov.io/gh/JuliaLang/julia/commits

https://coveralls.io/builds/56686154

vtjnash commented:
This just happened to be the commit where we fixed the coverage computation (https://github.com/JuliaCI/julia-buildkite/issues/188)
